pytorch_lightning/core/datamodule.py

```diff
@@ -398,3 +325,62 @@ def test_dataloader():
         if test_dataset is not None:
             datamodule.test_dataloader = test_dataloader
         return datamodule
+
+    def __new__(cls, *args: Any, **kwargs: Any) -> 'LightningDataModule':
+        obj = super(LightningDataModule, cls).__new__(cls)
+        # save `args` and `kwargs` for `__reduce__`
+        obj.__args = args
+        obj.__kwargs = kwargs
+        # track `DataHooks` calls and run `prepare_data` only on rank zero
+        obj.prepare_data = cls._track_data_hook_calls(obj, rank_zero_only(obj.prepare_data))
+        obj.setup = cls._track_data_hook_calls(obj, obj.setup)
+        obj.teardown = cls._track_data_hook_calls(obj, obj.teardown)
+        return obj
+
+    @staticmethod
+    def _track_data_hook_calls(obj: 'LightningDataModule', fn: callable) -> callable:
```

Comment:
Can we easily move the function here? I doubt that there will be user code, but you never know. So maybe we should have an alias with deprecation warning to ensure backwards compatibility?

Reply:
GitHub code search does not show a single occurrence:

https://github.com/search?q=track_data_hook_calls&type=code

I have no idea why this wasn't protected initially, it is a very internal piece of code.

An alias would not work as it now takes an `obj` instance too. To keep BC I would have to keep the old copy with `pragma: no-cover`.

I don't think it's worth it.